### ISSUE TITLE
Getting `nan` while training

### ISSUE THREAD

**Issue by sleepymalc** [ds4dm/ecole#324]
## Describe the bug
I followed the [`branching-imitation/example.ipynb`](https://github.com/ds4dm/ecole/blob/master/examples/branching-imitation/example.ipynb) while creating my own data instances. And I get `nan` while training.

## Setting
- OS: Linux/macOS
- Compiler: /
- Python version: 3.8
- Ecole version: 0.8

I want to use this to train on TSP, and I successfully produce the desired dataset/instances. But now, it seems like, after some iterations (less than 10), the loss becomes `nan`. I try to solve this by lowering the learning rate, but it doesn't help. And I have checked the document for `torch.nn.functional.cross_entropy()`, which said the input (in this case, `logits`) can be un-normalized, so I haven't tried to normalize it. Are there any other possible tips that may help solve this problem? The code snippets which cause the problem (I think) are as follows:
```python
def process(policy, data_loader, top_k=[1, 3, 5, 10], optimizer=None):
    mean_loss = 0
    mean_kacc = np.zeros(len(top_k))
    mean_entropy = 0

    n_samples_processed = 0
    with torch.set_grad_enabled(optimizer is not None):
        for batch in data_loader:
            batch = batch.to(device)

           ######## potentially problematic part starts

            logits = policy(batch.constraint_features, batch.edge_index, batch.edge_attr, batch.variable_features)
            logits = pad_tensor(logits[batch.candidates], batch.nb_candidates)

            cross_entropy_loss = F.cross_entropy(logits, batch.candidate_choices, reduction='mean')

            print("cross entropy loss", cross_entropy_loss)

            entropy = (-F.softmax(logits, dim=-1) * F.log_softmax(logits, dim=-1)).sum(-1).mean()
            loss = cross_entropy_loss - entropy_bonus * entropy

           ######## potentially problematic part ends

            if optimizer is not None:
                optimizer.zero_grad()
                loss.backward()
                optimizer.step()

            true_scores = pad_tensor(batch.candidate_scores, batch.nb_candidates)
            true_bestscore = true_scores.max(dim=-1, keepdims=True).values

            kacc = []
            for k in top_k:
                if logits.size()[-1] < k:
                    kacc.append(1.0)
                    continue
                pred_top_k = logits.topk(k).indices
                pred_top_k_true_scores = true_scores.gather(-1, pred_top_k)
                accuracy = (pred_top_k_true_scores == true_bestscore).any(dim=-1).float().mean().item()
                kacc.append(accuracy)
            kacc = np.asarray(kacc)

            mean_loss += cross_entropy_loss.item() * batch.num_graphs
            mean_entropy += entropy.item() * batch.num_graphs
            mean_kacc += kacc * batch.num_graphs
            n_samples_processed += batch.num_graphs

    mean_loss /= n_samples_processed
    mean_kacc /= n_samples_processed
    mean_entropy /= n_samples_processed
    return mean_loss, mean_kacc, mean_entropy
```
And the output is as follows
![image](https://user-images.githubusercontent.com/60427058/163762572-20d721ea-7f8b-4374-af3a-071e1cc12d64.png)
where I print the following three things:
![image](https://user-images.githubusercontent.com/60427058/163762624-3f26ff95-6b87-4862-b8fa-aabbf1cc3f28.png)
It seems like the `torch.nn.functional.cross_entropy()` first becomes `nan`, and then on the next iteration, everything becomes `nan` due to one bad update. Any helps is appreciated!

p.s. I'm not entirely sure whether this information is useful, but I'm working on TSP where I model this into lp, which makes the lp size pretty big. For an instance with 15 nodes, it works fine in 7 epochs, and if I simply **skip** the problematic instance (if the entropy is `nan`, just skip that instance), then it can be trained successfully (getting 100 epochs with acc@1 boosts from 33% to 66%). But if I use the same trick, then for tsp25, everything is skipped... So I think this may relate to the problem size?

**Comment by sleepymalc** [comment]
After tracking down the problem more thoroughly, I found that the problem is caused by samples. In every `sample.pkl`, there is `nan` in it. Is it normal? It's something like this:
```txt
{'episode': 0, 'instance': 'data/tsp15/instances/train/instance_9191.lp', 'seed': 3048896907, 'data': [(array([[-0.26726124, -0.31207214,  1.        , -0.0083808 ,  0.        ],
       [ 0.26726124,  0.31207214,  1.        ,  0.0083808 ,  0.        ],
       [-0.26726124, -0.22340463,  1.        , -0.00147708,  0.        ],
       ...,
       [-0.57735027, -0.        ,  1.        ,  0.        ,  0.03125   ],
       [ 0.57735027,  0.        ,  1.        , -0.        ,  0.03125   ],
       [ 0.70710678,  0.04004639,  1.        , -0.0070933 ,  0.        ]]), (array([[  0,   0,   0, ..., 451, 452, 452],
       [ 10,  33,  53, ..., 449, 107, 174]], dtype=uint64), array([-1., -1., -1., ..., -1.,  1.,  1.])), array([[0.01565678, 1.        , 0.        , ..., 0.        , 0.        ,
        0.        ],
       [0.08539868, 1.        , 0.        , ..., 0.        , 0.        ,
        0.        ],
       [0.06145908, 1.        , 0.        , ..., 0.        , 0.        ,
        0.        ],
       ...,
       [0.        , 0.        , 0.        , ..., 0.        , 0.        ,
        0.        ],
       [0.        , 0.        , 0.        , ..., 1.        , 0.        ,
        0.        ],
       [0.        , 0.        , 0.        , ..., 0.        , 0.        ,
        0.        ]])), 174, array([ 49, 114, 182, 202, 120, 145,  10,  74, 108, 179, 119,  53, 112,
       193, 135,   5,   2, 160, 122, 124,  31,  98,  82, 151,  30, 164,
        68,  80, 140, 133,  25, 197, 159,  65,  84, 180,   4, 171,  33,
        88, 200,  87, 175, 184, 115,  81,  67, 110, 125, 118, 194, 168,
       192,  90, 201, 101, 158,  59, 196,  32, 198,  15, 185, 103,   0,
       109, 163,  69,  22, 126,  58,  57, 144,  70, 204, 206, 139,  76,
       157,  60, 121,  19, 141, 117, 152,  37, 106, 130, 105, 138,  83,
        56, 188, 136,  21, 137, 174,  27, 102,  36,  24,  73, 111, 176,
        62, 169, 150,  75, 203, 104,  28, 199,  85,   1,  20,  79, 178,
       131,  61,   6, 142,  18,  97,  86, 161,  17, 162, 205,  78,  13,
        89, 166, 107,  34,  50,  14, 134,  77,  41,  91, 127,  45, 190,
        11, 129, 208,  35, 148,  43, 191, 143, 187,  23, 186,  72, 195,
       183, 128,  26, 165, 149,  92,   8,  51, 123, 147, 154,  95, 181,
        12,  16,  99,  54, 172,   9,  71,  52,  96, 132, 209],
      dtype=uint64), array([6.76038006e-06, 8.75337028e-05, 6.78716448e-05,            nan,
       7.14598685e-05, 9.57358432e-05, 4.21381082e-05,            nan,
       5.72917038e-06, 1.61069654e-04, 3.80939016e-05, 3.33135436e-05,
       2.76695778e-06, 8.75337028e-05, 6.96037578e-05, 1.97679543e-05,
       6.66692668e-05, 5.72917038e-06, 6.80109955e-05, 7.57969603e-05,
       1.26848053e-05, 3.29920649e-05, 1.83235866e-05, 1.25304530e+01,
       1.97679543e-05, 5.27784161e-05, 4.30842903e-05, 1.09451490e-05,
       8.57761256e-05,            nan, 1.26696561e-05, 9.52659453e-05,
       4.68368704e-05, 9.53442231e-05, 4.24346265e-05, 1.01577156e-05,
       5.05153568e-05, 4.04412536e-05,            nan,            nan,
                  nan, 1.02167271e-05,            nan, 3.39357548e+00,
                  nan, 6.83023225e-05,            nan,            nan,
                  nan, 1.01577156e-05, 7.03582946e-05, 3.79157885e-05,
       9.42153267e-05, 6.97607004e-05, 9.03699891e-05,            nan,
       5.48150923e-05, 1.25304530e+01, 1.40706657e-05, 1.09451490e-05,
       9.60717779e-05, 7.57446224e-05, 3.46173639e-05,            nan,
                  nan, 6.14728951e-05,            nan, 3.49924564e-05,
       8.19068384e-05, 6.92869734e-05, 3.49924564e-05, 3.80939016e-05,
       4.24540707e-05, 7.02702704e-05, 1.01577156e-05, 5.27784161e-05,
       7.57446224e-05, 1.27402211e-05, 4.34287671e-05, 7.19489382e-05,
       9.60915353e-06, 3.80939016e-05, 3.83527639e-05, 1.32762077e-04,
       5.94324436e-05, 4.87498052e-05, 4.21349466e-05, 6.83023225e-05,
       1.01577156e-05, 1.31578766e-04, 3.79157885e-05, 9.63983821e-05,
       7.12886014e-05,            nan,            nan, 6.97158392e-05,
       9.42289628e-05, 1.24239976e-04, 4.00380851e-05, 1.01493073e-04,
                  nan, 6.65692072e-05, 2.25138529e-05, 4.32310828e-05,
       1.01577156e-05, 4.34287671e-05, 3.97499571e-05, 1.03179186e+02,
       6.66954893e-05, 7.12886014e-05, 6.38002157e-05, 1.03381431e-04,
       3.58909833e-05,            nan, 3.39357548e+00, 1.04451314e-04,
                  nan, 7.09003234e-05, 4.85104857e-05, 1.27402211e-05,
       1.27402211e-05, 1.25339294e-04, 1.00554478e-04, 9.00820455e-05,
       9.58502183e-05, 3.80666934e-05, 6.10117510e-05, 1.26696561e-05,
       4.24540707e-05, 5.62963298e-05, 8.77682230e-05, 4.87498052e-05,
       9.77100886e-05, 3.82585363e-05, 2.76695778e-06, 7.98560379e-05,
       5.48123596e-05, 3.09619845e-05, 4.51780387e-05, 6.88488940e-05,
       6.97158392e-05, 1.31622571e-05, 5.48150923e-05, 4.22643570e-05,
       9.61969873e-05, 5.16031912e-05,            nan, 1.83235866e-05,
       5.62963298e-05, 6.76038006e-06, 8.48058525e-05, 3.97672814e-05,
       7.66140973e-05,            nan, 5.16031912e-05,            nan,
                  nan, 4.21381082e-05, 6.87731418e-05, 7.07406545e-05,
       9.22213573e-05, 8.32063541e-05, 5.67694521e-05, 7.06150755e-05,
       6.77161571e-05, 6.46435170e-05, 3.09619845e-05,            nan,
       6.58776753e-05, 1.08713856e-04,            nan, 2.25138529e-05,
       1.03966504e-04,            nan, 1.03179186e+02, 2.76695778e-06,
       8.14730330e-05,            nan, 1.01577156e-05, 5.00817603e-05,
       2.76695778e-06, 3.46173639e-05, 3.80939016e-05, 1.26848053e-05,
       7.33672949e-05, 7.57969603e-05, 3.83527639e-05, 4.18103421e-05,
       1.27402211e-05,            nan, 1.31622571e-05, 4.68368704e-05,
       9.60915353e-06, 6.96037578e-05, 3.55059713e-05, 7.32323575e-05,
       3.80939016e-05, 9.43546775e-05, 5.48123596e-05, 1.02167271e-05,
       6.92869734e-05, 1.04048696e-04, 5.20074515e-05, 6.66954893e-05,
       8.20214088e-05, 8.77682230e-05, 9.77898060e-05,            nan,
       1.48702340e-05, 1.53032446e-04,            nan,            nan,
                  nan,            nan,            nan,            nan,
                  nan,            nan,            nan,            nan,
                  nan,            nan,            nan,            nan,
                  nan,            nan,            nan,            nan,
                  nan,            nan,            nan,            nan,
                  nan,            nan,            nan,            nan,
                  nan,            nan,            nan,            nan,
                  nan,            nan,            nan,            nan,
                  nan,            nan,            nan,            nan,
                  nan,            nan,            nan,            nan,
                  nan,            nan,            nan,            nan,
                  nan,            nan,            nan,            nan,
                  nan,            nan,            nan,            nan,
                  nan,            nan,            nan,            nan,
                  nan,            nan,            nan,            nan,
                  nan,            nan,            nan,            nan,
                  nan,            nan,            nan,            nan,
                  nan,            nan,            nan,            nan,
                  nan,            nan,            nan,            nan,
                  nan,            nan,            nan,            nan,
                  nan,            nan,            nan,            nan,
                  nan,            nan,            nan,            nan,
                  nan,            nan,            nan,            nan,
                  nan,            nan,            nan,            nan,
                  nan,            nan,            nan,            nan,
                  nan,            nan,            nan,            nan,
                  nan,            nan,            nan,            nan,
                  nan,            nan,            nan,            nan,
                  nan,            nan,            nan,            nan,
                  nan,            nan,            nan,            nan,
                  nan,            nan,            nan,            nan,
                  nan,            nan,            nan,            nan,
                  nan,            nan,            nan,            nan,
                  nan,            nan,            nan,            nan,
                  nan,            nan,            nan,            nan,
                  nan,            nan,            nan,            nan,
                  nan,            nan,            nan,            nan,
                  nan,            nan,            nan,            nan,
                  nan,            nan,            nan,            nan,
                  nan,            nan,            nan,            nan,
                  nan,            nan,            nan,            nan,
                  nan,            nan,            nan,            nan,
                  nan,            nan,            nan,            nan,
                  nan,            nan,            nan,            nan,
                  nan,            nan,            nan,            nan,
                  nan,            nan,            nan,            nan,
                  nan,            nan,            nan,            nan,
                  nan,            nan,            nan,            nan,
                  nan,            nan,            nan,            nan,
                  nan,            nan,            nan,            nan,
                  nan,            nan,            nan,            nan,
                  nan,            nan,            nan,            nan,
                  nan,            nan,            nan,            nan,
                  nan,            nan,            nan,            nan,
                  nan,            nan,            nan,            nan,
                  nan,            nan,            nan,            nan,
                  nan,            nan,            nan,            nan,
                  nan,            nan,            nan,            nan,
                  nan,            nan,            nan,            nan,
                  nan,            nan,            nan,            nan,
                  nan,            nan,            nan,            nan,
                  nan,            nan,            nan,            nan,
                  nan,            nan,            nan,            nan,
                  nan,            nan,            nan,            nan,
                  nan,            nan,            nan,            nan,
                  nan,            nan,            nan,            nan,
                  nan,            nan,            nan,            nan,
                  nan,            nan,            nan,            nan,
                  nan,            nan,            nan,            nan,
                  nan,            nan,            nan,            nan,
                  nan,            nan,            nan,            nan,
                  nan,            nan,            nan,            nan,
                  nan,            nan,            nan,            nan,
                  nan,            nan,            nan,            nan,
                  nan,            nan,            nan,            nan,
                  nan,            nan,            nan,            nan,
                  nan,            nan,            nan,            nan,
                  nan,            nan,            nan,            nan,
                  nan,            nan,            nan,            nan,
                  nan,            nan,            nan,            nan,
                  nan,            nan,            nan,            nan,
                  nan,            nan,            nan,            nan,
                  nan,            nan,            nan,            nan,
                  nan,            nan,            nan,            nan,
                  nan,            nan,            nan,            nan,
                  nan,            nan,            nan,            nan,
                  nan,            nan,            nan,            nan,
                  nan,            nan,            nan,            nan,
                  nan,            nan,            nan,            nan,
                  nan,            nan,            nan,            nan,
                  nan,            nan,            nan,            nan,
                  nan,            nan,            nan,            nan,
                  nan,            nan,            nan,            nan,
                  nan,            nan,            nan,            nan,
                  nan,            nan,            nan,            nan,
                  nan,            nan,            nan,            nan,
                  nan,            nan,            nan,            nan,
                  nan,            nan,            nan,            nan,
                  nan,            nan,            nan,            nan,
                  nan,            nan,            nan,            nan,
                  nan,            nan,            nan,            nan,
                  nan])]}
```
